py3status/modules/gazelle.py

```diff
@@ -0,0 +1,171 @@
+
+    def _login(self):
+        self.py3.request(
+                self.baseurl+"/login.php",
+                data={'username': self.username, 'password': self.password},
+                headers={'User-Agent': 'Mozilla/5.0'},
+                cookiejar=self._cookiejar
+                )
+        try:
+            self._index()  # just to try if login info are good
+        except RequestException:
+            self.py3.log(
+                    "Cannot login to %s" % self.baseurl,
+                    level=self.py3.LOG_ERROR
+                    )
+            raise
+        try:
+            self._cookiejar.save(ignore_discard=True)
+        except IOError:
+            self.py3.log(
+                    "Cannot write cookie at %s" % self.cookiepath,
+                    level=self.py3.LOG_WARNING
+                    )
+        except AttributeError:
+            pass  # That means that no cookie pass has been set
+        else:
+            self.py3.log(
+                    "Created cookie at %s" % self.cookiepath,
+                    level=self.py3.LOG_INFO
+                    )
+
+    def _index(self):
+        resp = self.py3.request(
+                self.baseurl+"/ajax.php",
+                params={'action': "index"},
+                headers={
+                    'User-Agent': 'Mozilla/5.0'
+                    },
+                cookiejar=self._cookiejar
+                )
+        return resp.json()
+
+    def post_config_hook(self):
+        if self.cookiepath:
+            self._cookiejar = MozillaCookieJar(self.cookiepath)
+            try:
+                self._cookiejar.load(ignore_discard=True)
+                self._index()  # just to try if saved cookie is good
+            except IOError:
+                self.py3.log(
+                        "Cannot read cookie at %s" % self.cookiepath,
+                        level=self.py3.LOG_INFO
+                        )
+                self._login()
+            except RequestException:
+                self.py3.log(
+                        "Cookie invalid or outdated. Re-logging.",
+                        level=self.py3.LOG_INFO
+                        )
+                self._login()
+        else:
+            self._cookiejar = CookieJar()
+
+    def gazelle(self):
+        index = self._index()
+        if index['status'] != "success":
+            msg = "Tracker failed to give me user info"
+            self.py3.log(msg, level=self.py3.LOG_ERROR)
+            raise RequestException(msg)
+        result = index['response']
+        data = {
+            'ratio': result['userstats']['ratio'],
+            'requiredratio': result['userstats']['requiredratio'],
+            'download': self._readable_size(result['userstats']['downloaded']),
+            'upload': self._readable_size(result['userstats']['uploaded']),
+            'messages': result['notifications']['messages'],
+            'notifs': result['notifications']['notifications']
```

Comment:
This seems like you're cherry-picking some data. Is there more data? Maybe we should use `flatten_dict` and expose everything? We can get more new placeholders without changing anything. Can you output `index`? :-)

Also, what's the difference between `messages` and `notifications`? I think with more work, we could add `urgent` for them... and if `index` contains an url, we might be able to use that too for `on_click` clicking purposes. ;-)

The new placeholders might look like this.
```
{userstats_requiredratio}
{userstats_downloaded}
{userstats_uploaded}
{notifications_messages}
{notifications_notifications}
etc...
```
If you don't like that, we might be able to skip prefixes too if we can guarantee no clashes.

I have been doing the good job of converting few cherrypicked placeholders to exposing all possible placeholders in lot of modules, eg cmus, moc, deadbeef, nvidia_smi, coin_market, and more... so I'd like for this module to do the same thing too.... It can be useless information, but still good information somehow.

Reply:
See [here](https://github.com/WhatCD/Gazelle/wiki/JSON-API-Documentation#index) for a list of mandatory json terms prefent in the response. I'll see about that, dunno if that's a good idea though.

Reply:
In fact, can't do here: some values in the json are boolean and you can't give booleans in the py3status response data, right?

Reply:
We can deal with boolean. New module `input` contains over 100 boolean (keys).

See `hueshift` for a good example. There are two boolean in the `format`. One will be used to disable the module when we detect `redshift` process via `pgrep` and the other one is to determine when to hide `is_scrolling`.

Not only that, but we can do some conditions too like...
* `'\?if=percent>50 More than halfway.'`
* `'\?if=percent>95 Almost done!'`

Reply:
Knowing you made your own `cmus` module long time ago... The `cmus` I made have this format too.
```
    format: display format for this module
        *(default '[\?if=is_started [\?if=is_playing > ][\?if=is_paused \|\| ]
        [\?if=is_stopped .. ][[{artist}][\?soft  - ][{title}]
        |\?show cmus: waiting for user input]]')*
```

Reply:
I dodn't say you couldn't use booleans in `format` field. I said that they can't be used as placeholder (which may only be strings, integers or floats if I'm not wrong) and so they can't be passed in the `full_text` data dict.

Reply:
They can be used as placeholders. It'd show `True` or `False`. Useless. In this case, we almost always will want to use `\?if=placeholder_name` instead. Also, check out the `Control placeholders` for module `moc` and `cmus`. All of that is boolean.

Reply:
Also on another topic, I wonder why cmus got finally it's own module whereas back in the days it was refused because it was preferable to adapt the existing music module to support cmus in order to avoid creating modules for specific program instead of modules for specific functions. I understood and agreed with this principle.

Also now cmus supports MPRIS wich wasn't the case back then, which makes it entirely compatible with the mpris module.

But we might talk about that in another place.

Reply:
OK then, never mind about placeholders. I'll think about that.

Reply:
Don't misunderstand, I'm not angry or anything, especially not against you, I'm just curious about how the rules changed and I want to understand those rules in order to make my PRs in this objective and thus spend much less time on fixing philosophical stuff.

@ultrabug @tobes the documentation contains all the technical knowledge to create a well functioning module, and that's great. Maybe you should also consider writing a contribution manifest explaining for instance which conditions are needed in order to create a new module, your vision of module improvement, what is the general direction of the project, what are the good practices...
I know that a little of this info is spread in the whole *writing custom py3status modules* section, but I really think that those "philosophical" considerations should have their own subsection, or even their own entire section.

Reply:
I'll try and explain this the best I can.

I think @tobes wanted us to go in a direction where we don't script anything in the code by exposing everything (if possible) as placeholders then users can manipulate the `if`, `else`, etc in the `format` instead of making pull requests. See issue... https://github.com/ultrabug/py3status/issues/752

If you want to, try `window_manipulation`.... That module contains an amount of placeholders and I was able to manipulate lot of things.... With more placeholders, we can do more neat things... without touching the code.

The key is essentially to get all placeholders we can get our hands on in all modules... and have @tobes become a farmer where he can grow the formatters this summer.

Users want colors. Users don't want colors. We can make both things possible.
Users want urgent. Users don't want urgent. We can make both things possible.
Users want floats or ints... We can make both things possible.
Users want to hide or scrub things. We're going to try and make that possible.
Users... et cetera.

This requires lot of changes though... For instance, all modules should stop returning `"color": color` because `module { color = '#ffccff'}` wouldn't work with that since it's hardcoded... Like I said, lot of changes.
* Remove hardcoded color returns.
* Merge separate formats into one `format`. 
* Undo any `int(floating_value)`.
* How to deprecate things properly? I hate this.
* We can compare basic words. `\?if=class=VIP Good Boy!`
* Remove configs like `hide_if_zero` in favor of `'\?if=apt {apt} update}'` or `'\?not_zero {apt} update'`. 

Users does not know about those things too... due to lack of docs and whatnot. Users may not understand that amount of customization too... so I have been trying to come up with decent examples for users to copy and paste into their formats... and make small changes from there.

One problem I think is that @ultrabug may be wanting to keep things stable and make it easy for users to change settings via configs instead of scriptable `format`.

Another problem, I make many PRs. It scares and/or annoys them.
Another problem, they're slow on reviewing and/or merging. It annoys me.

Disclaimer: There are lot of things I don't know about py3status or python... and most of time, I'm just guessing things... and how things work around here... and I'm always doing something wrong too. I'm just a contributor here. Same as you.

Reply:
Thanks for your explanations. I still think that it must be formalized somewhere "officially" but this is the idea.

By the way, I find this format language very uncomfortable and unintuitive, but I guess it cannot be helped and this is the best that could have been done. Also this would be so much easier to use **if this was documented somewhere**. People can't think to use stuff it they are not documented. I don't know how long ago this feature exists (I don't remind such a thing from back in the days) and how long ago the `color` field method has been deprecated, but **it's urgent and capital to write something about it in the official documentation**. Else casual contributors like me will continue coding with their feet and everybody will spend more time on debugging stupidities, which leads to less time to important thing like documentation and "good" coding, which is a vicious circle.

I know you're also a (much less) casual contributor (than me) so that's why it needs to be formalized and acknowledged by an official core developer.

I'm not mad, I just express my feeling about what I objectively think to be the good way to progress, so please excuse my strong sentences with **bold text**, if it offend someone that surely wasn't my intention.

Reply:
@raspbeguy 

> I just express my feeling about what I objectively think to be the good way to progress, 

Expressing your opinions is always encouraged here.  And both you and @lasers raise some good points.  There are definitely some things that need improving with the project.  Some documentation is desperately needed such as the formatting and some better guidance on module development.

From my point of view I want things to be easy for user and I think we should develop for 2 distinct users.

a) Users who just want things to work.  To me a good example would be the `volume_status` as a user I should be able to just do `order += 'volume_status'` and should get a sensible working volume control (I don't think we are here as the volume changing controls isn't by default)

b) expert users who want to set things up exactly as they want full control of colors etc.  This also to an extent includes people who just want to make small changes say to the format.

Because of this I want to keep things like default formats really simple.  @lasers for example likes to make lots of format usage (different colors for different states)  I resist all attempts for complex default formats (@ultrabug is more forgiving here)

for me simplicity is key so I will always struggle with things like

```
-    format = '{caps} {num} {scr}'

+    format = ('[\?if=caps_lock&color=good CAPS|\?color=bad CAPS] '
+              '[\?if=num_lock&color=good NUM|\?color=bad NUM] '
+              '[\?if=scroll_lock&color=good SCR|\?color=bad SCR]')
```

mainly because to me it shows we need to improve the formatter to make this simpler

I'm also keen to have as few modules as we can that do the same thing (music players being an example)  I'd much prefer one really solid mpris module rather than many different modules one for each player.

@lasers 
>  I make many PRs. It scares and/or annoys them.

I'd say that it is more it makes it harder to manage things as reviewing all open PRs is not easily possible when there are so many of them.  Things can get lost etc.  Sometimes things are just slow.  You also like to make many changes in a PR and so a small part can easily block the entire PR.  I know this is frustrating for you.

Reply:
I must agree with @tobes on the "keep configuration simplicity" thing. That format language is dreadful, and a lot of regular user won't even try to understand it. Also, I have no alternative to propose, but I know this will be a real pain for me to use that.

Reply:
@tobes Maybe that contribution manifest I was speaking about should include a line like "Every new core feature should be documented in the official documentation" so that a PR adding features to core code (not a module) shouldn't be accepted until the contributor (or anybody) takes the time to write in the official doc what this new feature is and how to use it. I had the case with the Request code that contained a lot of non-documented stuff so I wasn't even able to write down my own feature documentation.

I had plans to code stuff, but I guess I'll spend a few PR to write some documentation (about what I can understand in the code I touched).